File: lista-de-exercicios/exercicio-3/Jokenpo.ts
export enum Escolhas {PEDRA, PAPEL, TESOURA};

export interface JogadorBase {
    getEscolha(): Escolhas;
}

export class Jokenpo {
    private jogador1: JogadorBase;
    private jogador2: JogadorBase;

    public constructor (jogador1: JogadorBase, jogador2: JogadorBase) {
        this.jogador1 = jogador1;
        this.jogador2 = jogador2;
    }

    public jogar(): JogadorBase | string {
        const jogador1 = this.jogador1.getEscolha();
        const jogador2 = this.jogador2.getEscolha();
        const pedra = Escolhas.PEDRA;
        const papel = Escolhas.PAPEL;
        const tesoura = Escolhas.TESOURA;
        
        if (jogador1 == jogador2)
            return 'Empate';
        else if (jogador1 == papel && jogador2 == pedra)
            return `Vencedor: Player 1 com ${Escolhas[jogador1]}`;
        else if (jogador1 == tesoura && jogador2 == papel)
            return `Vencedor: Player 1 com ${Escolhas[jogador1]}`;
        else if (jogador1 == pedra && jogador2 == tesoura)
            return `Vencedor: Player 1 com ${Escolhas[jogador1]}`;
        return `Vencedor: Player 2 com ${Escolhas[jogador2]}`;
    }
}

File: lista-de-exercicios-java/src/main/java/org/example/exercicio03/Jokenpo.java
package org.example.exercicio03;

import static org.example.exercicio03.Escolhas.PEDRA;
import static org.example.exercicio03.Escolhas.PAPEL;
import static org.example.exercicio03.Escolhas.TESOURA;

public class Jokenpo {
    private JogadorBase jogador1;
    private JogadorBase jogador2;

    public Jokenpo(JogadorBase jogador1, JogadorBase jogador2) {
        this.jogador1 = jogador1;
        this.jogador2 = jogador2;
    }

    public String jogar() {
        Escolhas jogada1 = this.jogador1.getEscolha();
        Escolhas jogada2 = this.jogador2.getEscolha();

        if(jogada1 == jogada2)
            return "Empate";
        else if(jogada1 == PAPEL && jogada2 == PEDRA)
            return "Vencedor: Player 1 com " + jogada1.toString();
        else if(jogada1 == TESOURA && jogada2 == PAPEL)
            return "Vencedor: Player 1 com " + jogada1.toString();
        else if(jogada1 == PEDRA && jogada2 == TESOURA)
            return "Vencedor: Player 1 com " + jogada1.toString();
        return "Vencedor: Player 2 com " + jogada2.toString();
    }
}
